include/nbl/asset/IRayTracingPipeline.h

```diff
@@ -0,0 +1,171 @@
+#ifndef _NBL_ASSET_I_RAY_TRACING_PIPELINE_H_INCLUDED_
+#define _NBL_ASSET_I_RAY_TRACING_PIPELINE_H_INCLUDED_
+
+#include "nbl/asset/IShader.h"
+#include "nbl/asset/IPipeline.h"
+
+#include <span>
+
+
+namespace nbl::asset
+{
+
+class IRayTracingPipelineBase : public virtual core::IReferenceCounted
+{
+  public:
+    struct SShaderGroupsParams
+    {
+      constexpr static inline uint32_t ShaderUnused = 0xffFFffFFu;
+
+      struct SGeneralShaderGroup
+      {
+        uint32_t shaderIndex = ShaderUnused;
+      };
+
+      struct SHitShaderGroup
+      {
+        uint32_t closestHitShaderIndex = ShaderUnused;
+        uint32_t anyHitShaderIndex = ShaderUnused;
+        uint32_t intersectionShaderIndex = ShaderUnused;
+      };
+
+      SGeneralShaderGroup raygenGroup;
+      std::span<SGeneralShaderGroup> missGroups;
+      std::span<SHitShaderGroup> hitGroups;
+      std::span<SGeneralShaderGroup> callableGroups;
+
+      inline uint32_t getShaderGroupCount() const
+      {
+        return 1 + hitGroups.size() + missGroups.size() + callableGroups.size();
+      }
+
+    };
+    using SGeneralShaderGroup = SShaderGroupsParams::SGeneralShaderGroup;
+    using SHitShaderGroup = SShaderGroupsParams::SHitShaderGroup;
+    struct SCachedCreationParams final
+    {
+      uint32_t maxRecursionDepth;
+    };
+};
+
+template<typename PipelineLayoutType, typename ShaderType>
+class IRayTracingPipeline : public IPipeline<PipelineLayoutType>, public IRayTracingPipelineBase
+{
+  public:
+
+    using SGeneralShaderGroupContainer = core::smart_refctd_dynamic_array<SGeneralShaderGroup>;
+    using SHitShaderGroupContainer = core::smart_refctd_dynamic_array<SHitShaderGroup>;
+
+    struct SCreationParams : IPipeline<PipelineLayoutType>::SCreationParams
+    {
+      protected:
+        using SpecInfo = ShaderType::SSpecInfo;
+        template<typename ExtraLambda>
+        inline bool impl_valid(ExtraLambda&& extra) const
+        {
+          if (!IPipeline<PipelineLayoutType>::SCreationParams::layout)
+            return false;
+
+          core::bitflag<ICPUShader::E_SHADER_STAGE> stagePresence = {};
+          for (const auto info : shaders)
+            if (info.shader)
+            {
+              if (!extra(info))
+                return false;
+              const auto stage = info.shader->getStage();
+              if (stage > ICPUShader::E_SHADER_STAGE::ESS_CALLABLE || stage < ICPUShader::E_SHADER_STAGE::ESS_RAYGEN)
+                return false;
+              if (stage == ICPUShader::E_SHADER_STAGE::ESS_RAYGEN && stagePresence.hasFlags(hlsl::ESS_RAYGEN))
+                return false;
```

Comment:
would be nice to leave a comment "only one raygen allowed"

Reply:
hmm actually if Vulkan spec doesn't mandate this, I'd allow multiple raygen shaders in the span (after all you only have one `raygenGroup` later)